Signal/src/view controllers/SignalAttachment.swift

```diff
@@ -0,0 +1,477 @@
+//
+//  Copyright (c) 2017 Open Whisper Systems. All rights reserved.
+//
+
+import Foundation
+import MobileCoreServices
+
+enum SignalAttachmentError: String {
+    case fileSizeTooLarge
+    case invalidData
+    case couldNotParseImage
+    case couldNotConvertToJpeg
+    case invalidFileFormat
+}
+
+enum TSImageQuality: Int {
+    case uncropped
+    case high
+    case medium
+    case low
+}
+
+// Represents a possible attachment to upload.
+// The attachment may be invalid.
+//
+// Signal attachments are subject to validation and 
+// in some cases, file format conversion.
+//
+// This class gathers that logic.  It offers factory methods
+// for attachments that do the necessary work. 
+//
+// The return value for the factory methods will be nil if the input is nil.
+//
+// [SignalAttachment hasError] will be true for non-valid attachments.
+//
+// TODO: Perhaps do conversion off the main thread?
+// TODO: Show error on error.
+// TODO: Show progress on upload.
+class SignalAttachment: NSObject {
+
+    static let TAG = "[SignalAttachment]"
+
+    // MARK: Properties
+
+    let data: Data!
+
+    // Attachment types are identified using UTIs.
+    //
+    // See: https://developer.apple.com/library/content/documentation/Miscellaneous/Reference/UTIRef/Articles/System-DeclaredUniformTypeIdentifiers.html
+    let dataUTI: String!
+
+    var error: SignalAttachmentError? {
+        didSet {
+            AssertIsOnMainThread()
+
+            assert(oldValue == nil)
+            Logger.verbose("\(SignalAttachment.TAG) Attachment has error: \(error)")
+        }
+    }
+
+    // To avoid redundant work of repeatedly compressing/uncompressing
+    // images, we cache the UIImage associated with this attachment if
+    // possible.
+    public var image: UIImage?
+
+    // MARK: Constants
+
+    /**
+     * Media Size constraints from Signal-Android
+     * (org/thoughtcrime/securesms/mms/PushMediaConstraints.java)
+     */
+    static let kMaxFileSizeGif = 5 * 1024 * 1024
+    static let kMaxFileSizeImage = 420 * 1024
+    static let kMaxFileSizeVideo = 100 * 1024 * 1024
+    static let kMaxFileSizeAudio = 100 * 1024 * 1024
+    // TODO: What should the max file size on "other" attachments be?
```

Comment:
You know, I went to https://github.com/WhisperSystems/Signal-Android/blob/master/src/org/thoughtcrime/securesms/mms/PushMediaConstraints.java and it turns out these values are out of sync with Android.  I've aligned them.